File: DemoScrollViewPageControl/DemoScrollViewPageControl/AViewController.m
//
//  AViewController.m
//  DemoScrollViewPageControl
//
//  Created by zj－db0465 on 16/1/13.
//  Copyright © 2016年 icetime17. All rights reserved.
//

#import "AViewController.h"

#import "AView.h"

@interface AViewController () <AViewScrollDelegate>

@end

@implementation AViewController

- (void)viewDidLoad {
    [super viewDidLoad];
    // Do any additional setup after loading the view.
    
    self.view.backgroundColor = [UIColor whiteColor];
    
    AView *aView = [[AView alloc] initWithFrame:CGRectMake(0, 100, self.view.frame.size.width, 300)];
    [self.view addSubview:aView];
    aView.scrollDelegate = self;
    
    [self addBackButton];
}

- (void)didReceiveMemoryWarning {
    [super didReceiveMemoryWarning];
    // Dispose of any resources that can be recreated.
}

#pragma mark - back button

- (void)addBackButton {
    UIButton *btn = [[UIButton alloc] initWithFrame:CGRectMake(0, self.view.frame.size.height - 100, self.view.frame.size.width, 50)];
    [btn setTitle:@"Back" forState:UIControlStateNormal];
    [btn setTitleColor:[UIColor blueColor] forState:UIControlStateNormal];
    [btn setTitleColor:[UIColor redColor] forState:UIControlStateHighlighted];
    [btn addTarget:self action:@selector(actionBack:) forControlEvents:UIControlEventTouchUpInside];
    btn.layer.borderColor = [UIColor redColor].CGColor;
    btn.layer.borderWidth = 2.0f;
    [self.view addSubview:btn];
}

- (void)actionBack:(UIButton *)sender {
    [self dismissViewControllerAnimated:YES completion:nil];
}

#pragma mark - AViewScrollDelegate

- (void)viewDidScroll {
    NSLog(@"AViewController viewDidScroll");
}

@end

File: DemoMLeaksFinder/DemoMLeaksFinder/AViewController.swift
//
//  AViewController.swift
//  DemoMLeaksFinder
//
//  Created by Chris Hu on 16/12/30.
//  Copyright © 2016年 com.icetime17. All rights reserved.
//

import UIKit

class AViewController: UIViewController {

    var timer: Timer!
    
    override func viewDidLoad() {
        super.viewDidLoad()
        
        view.backgroundColor = UIColor.gray
        
        let btnTest = UIButton(frame: CGRect(x: 0, y: 0, width: 100, height: 30))
        view.addSubview(btnTest)
        btnTest.center = view.center
        btnTest.setTitle("Timer", for: .normal)
        btnTest.addTarget(self, action: #selector(AViewController.actionBtnTest), for: .touchUpInside)
        
        let btnBack = UIButton(frame: CGRect(x: 50, y: 100, width: 100, height: 30))
        view.addSubview(btnBack)
        btnBack.setTitle("Back", for: .normal)
        btnBack.addTarget(self, action: #selector(AViewController.actionBtnBack), for: .touchUpInside)
    }
    
    override func viewWillDisappear(_ animated: Bool) {
        super.viewWillDisappear(animated)
        
        // 如果不释放timer，则会触发Memory警告
        timer.invalidate()
        timer = nil
    }
    
    func actionBtnTest() {
        timer = Timer.scheduledTimer(timeInterval: 1, target: self, selector: #selector(AViewController.actionTimer), userInfo: nil, repeats: true)
        timer.fire()
    }
    
    func actionBtnBack() {
        dismiss(animated: true, completion: nil)
    }
    
    func actionTimer() {
        print("timer running")
    }
    
}
